### PR TITLE
For evaluation: Implementation of MarkdownDeep.net over Journaley.

### PR DESCRIPTION
**Pull request by crxtrdude**
Replaced MarkdownSharp with [MarkdownDeep.net](https://github.com/toptensoftware/markdowndeep). Has better compatibility with some Markdown features that Day One supports including Markdown Extra features (tables and footnotes for one), although there are some certain things it doesn't have. That's where fixes comes in, the most important is for single line breaks.

For more information, refer to the commits for details.

### CODE REVIEW

File: Journaley/Forms/MainForm.cs [hunk 1]
```diff
@@ -638,26 +639,35 @@ private void OnGetMinMaxInfo(ref Message m)
         /// </summary>
         private void UpdateWebBrowser()
         {
+            string initialString = this.SelectedEntry.EntryText;
             string formattedString = string.Format(
-                    "<style type=\"text/css\">\n<!-- Font CSS -->\n{0}\n<!-- Size CSS -->\n{1}\n<!-- Custom CSS -->\n{2}\n</style><html><body><div>{3}</div></body></html>",
+                    "<style type=\"text/css\">\n<!-- Font CSS -->\n{0}\n<!-- Size CSS -->\n{1}\n<!-- Custom CSS -->\n{2}\n</style>\n<html>\n<body>\n<div>\n{3}\n</div>\n</body>\n</html>",
```
Comment (crxtrdude) [inline Journaley/Forms/MainForm.cs]: Added proper line breaks for the tags.

File: Journaley/Forms/MainForm.cs [hunk 2]
```diff
@@ -666,25 +676,52 @@ private string RemoveLineBreaksWithinLists(string formattedString)
...
-                builder.AppendLine(line);
+                    builder.AppendLine(newLineString);
+                    paragraphBuilder.Clear();
+                }
             }
 
             builder.AppendLine(lines.Last());
 
-            return builder.ToString();
+            strippedOver = Regex.Replace(builder.ToString(), @"^(\r\s)", string.Empty, RegexOptions.Multiline);
```
Comment (crxtrdude) [inline Journaley/Forms/MainForm.cs]: Did this because the output that the `builder` creates has a lot of very bad carriage returns, so I stripped them into empty. Tested this, did not affect the output and the HTML stayed intact.

File: Journaley.Core/Journaley.Core.csproj [hunk 3]
```diff
@@ -97,7 +97,9 @@
       <ErrorText>This project references NuGet package(s) that are missing on this computer. Enable NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is {0}.</ErrorText>
     </PropertyGroup>
     <Error Condition="!Exists('..\packages\StyleCop.MSBuild.4.7.48.2\build\StyleCop.MSBuild.Targets')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\StyleCop.MSBuild.4.7.48.2\build\StyleCop.MSBuild.Targets'))" />
+    <Error Condition="!Exists('$(SolutionDir)\.nuget\NuGet.targets')" Text="$([System.String]::Format('$(ErrorText)', '$(SolutionDir)\.nuget\NuGet.targets'))" />
```
Comment (yyoon) [inline Journaley.Core/Journaley.Core.csproj]: Was there a particular reason to change this file?

File: Journaley/Forms/MainForm.cs [hunk 4]
```diff
@@ -475,7 +475,8 @@ private Markdown Markdown
                 if (this.markdown == null)
                 {
                     this.markdown = new Markdown();
-                    ((MarkdownOptions)this.markdown.Options).AutoNewLines = true;
+                    this.markdown.ExtraMode = true;
+                    this.markdown.SafeMode = false;
```
Comment (yyoon) [inline Journaley/Forms/MainForm.cs:479]: What are these two options?

Comment (crxtrdude) [inline Journaley/Forms/MainForm.cs:479]: From the [website](http://www.toptensoftware.com/markdowndeep/api):

> bool SafeMode
> Set to true to only allow whitelisted safe html tags
> 
> bool ExtraMode
> Set to true to enable ExtraMode, which enables the same set of features as implemented by PHP Markdown Extra.

Extramode is needed, becuase most of the implementations in Day One uses Markdown Extra stuff like tables and footnotes

File: Journaley/Forms/MainForm.cs [hunk 5]
```diff
@@ -638,53 +639,18 @@ private void OnGetMinMaxInfo(ref Message m)
         /// </summary>
         private void UpdateWebBrowser()
         {
+            string initialString = this.SelectedEntry.EntryText;
+            string docType = "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01//EN\" \"http://www.w3.org/TR/html4/strict.dtd\">";
```
Comment (yyoon) [inline Journaley/Forms/MainForm.cs]: I'm fine with adding the doctype declaration, but why not HTML5? Let's just go with:

``` csharp
string docType = "<!DOCTYPE html>";
```

File: Journaley.Core/Journaley.Core.csproj (hunk 3 above)
Comment (crxtrdude) [inline Journaley.Core/Journaley.Core.csproj]: I really don't know. I think this was when I enabled Nuget Package Recovery in VS.

File: Journaley/Forms/MainForm.cs (hunk 5 above)
Comment (crxtrdude) [inline Journaley/Forms/MainForm.cs]: ~~We consider the ones that have IE 8 or 9 where HTML5 might not work. Remember that the WebBrowser control relies on the Internet Explorer available. But yeah, we could put HTML5, but I'm not sure if it renders to other people's computers nicely.~~

On second thought, maybe we could use the HTML5 doctype since the other IEs will conform to it anyway.

File: Journaley/Forms/MainForm.cs (hunk 4 above)
Comment (yyoon) [inline Journaley/Forms/MainForm.cs:479]: Thanks. We might want to turn on the SafeMode option later for resolving #118. Let's keep this false for now.

File: Journaley/Journaley.csproj [hunk 6]
```diff
@@ -499,10 +501,12 @@
       <ErrorText>This project references NuGet package(s) that are missing on this computer. Enable NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is {0}.</ErrorText>
     </PropertyGroup>
     <Error Condition="!Exists('..\packages\StyleCop.MSBuild.4.7.48.2\build\StyleCop.MSBuild.Targets')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\StyleCop.MSBuild.4.7.48.2\build\StyleCop.MSBuild.Targets'))" />
+    <Error Condition="!Exists('$(SolutionDir)\.nuget\NuGet.targets')" Text="$([System.String]::Format('$(ErrorText)', '$(SolutionDir)\.nuget\NuGet.targets'))" />
```
Comment (yyoon) [inline Journaley/Journaley.csproj]: Please remove the NuGet package restore related lines (504 & 509).

File: Journaley.Core/Journaley.Core.csproj (hunk 3 above)
Comment (yyoon) [inline Journaley.Core/Journaley.Core.csproj]: No, you shouldn't enable that in the solution. In fact, it turns out these changes (in `Journaley.Core`, `Journaley`, `VersionExtractor`) broke the build on my side. Please remove these lines from all the affected projects.

You should just go to `Tools` -> `Options...` -> `NuGet Package Manager` -> `General` and check both the `Allow NuGet to download missing packages` and `Automatically check for missing pakcages during build in Visual Studio` options. The solution should build fine after enabling these options. If not, let's troubleshoot that together later on.

File: VersionExtractor/VersionExtractor.csproj [hunk 7]
```diff
@@ -11,6 +11,8 @@
     <AssemblyName>VersionExtractor</AssemblyName>
     <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
     <FileAlignment>512</FileAlignment>
+    <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\</SolutionDir>
```
Comment (yyoon) [inline VersionExtractor/VersionExtractor.csproj]: Please revert this entire file.

File: Journaley.Core/Journaley.Core.csproj (hunk 3 above)
Comment (crxtrdude) [inline Journaley.Core/Journaley.Core.csproj]: I'll reset my repo and I'll do this again. I'll message you if it worked with just removing the code first.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Helper class for fixing MarkdownDeep parsed HTML.
+    /// </summary>
+    public class PostMarkdownParser
+    {
+        /// <summary>
+        /// Perform fixes after MarkdownDeep parsing for publishing
+        /// to Journaley.
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs:17]: You need to be more explicit here and say what it fixes.
1. automatically add `<br />` tags where the user hit enter.
2. add strikethrough style support

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        /// <returns>Properly formatted HTML string for publishing.</returns>
+        public static string PostMarkdown(string formattedString)
+        {
+            // First convert all code blocks using <p> into <pre>
+            // which is used in almost all implementations of
+            // Markdown, even in the original Markdown specs.
+            formattedString = Regex.Replace(formattedString, @"<p><code>", "<pre><code>", RegexOptions.Multiline);
+            formattedString = Regex.Replace(formattedString, @"</code></p>", "</code></pre>", RegexOptions.Multiline);
+
+            System.Text.StringBuilder builder = new System.Text.StringBuilder();
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: Don't need the `System.Text.` part. You can just say `StringBuilder` because you already added `using System.Text` above.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            System.Text.StringBuilder paragraphBuilder = new System.Text.StringBuilder();
+
+            var isNewLine = 0;
+
+            string newLineString;
+            string strippedOverString;
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: `strippedOverString` is declared here and then not used until the end of the method. Either declare this when it's actually used, or remove this variable.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+            var lines = formattedString.Split(
+                new string[] { "\r\n", "\n" },
+                StringSplitOptions.None);
+
+            for (int i = 0; i < lines.Length - 1; ++i)
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: I think you're following the pattern I used for removing `<br />` tags before, but this loop is not appropriate here.

Just use `foreach (string line in lines)` instead, and remove `string line = lines[i];`, and remove `builder.AppendLine(lines.Last());` in line 108.

As it stands now, this loop doesn't run for the last line, and for example, if there was a strikethrough markup in the last line, it wouldn't be applied correctly.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+                }
+
+                if (isNewLine == 1)
+                {
+                    paragraphBuilder.AppendLine(line);
+
+                    if (line.Contains("</p>"))
+                    {
+                        // Single line strikethroughs.
+                        line = Regex.Replace(line, @"(~~)(.*?)(~~)", "<del>$2</del>", RegexOptions.Multiline);
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: Is the `RegexOptions.Multiline` option necessary? I think the line should always be a single line at this point.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+                    }
+                }
+                else if (isNewLine == 0)
+                {
+                    paragraphBuilder.AppendLine(line);
+
+                    newLineString = paragraphBuilder.ToString();
+
+                    // Multiline strikethrough.
+                    if (Regex.IsMatch(newLineString, @"(~~)(.*?)(~~)"))
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: Can you explain me why you need this `if` statement? Isn't it enough to just have the first `Regex.Replace` statement?

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+                    {
+                        newLineString = Regex.Replace(newLineString, @"(~~)(.*?)(~~)", "<del>$2</del>", RegexOptions.Multiline);
+                    }
+                    else
+                    {
+                        newLineString = Regex.Replace(newLineString, @"<p>(~~)", "<p><del>", RegexOptions.Multiline);
+                        newLineString = Regex.Replace(newLineString, @"(~~)</p>", "</del></p>", RegexOptions.Multiline);
+                    }
+
+                    newLineString = Regex.Replace(newLineString, @"^([\w\*\>\<\[][^\r\n]*)(?=\r?\n[\w\*\>\<\[].*$)", "$1<br />", RegexOptions.Multiline);
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: I can't really understand this regex. Can you explain me what it does?

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+                }
+                else
+                {
+                    // For <pre> tagged text, just skip check and append to builder
+                    builder.AppendLine(line);
+                }
+            }
+
+            builder.AppendLine(lines.Last());
+            strippedOverString = Regex.Replace(builder.ToString(), @"^(\r\s)", string.Empty, RegexOptions.Multiline);
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: I'm not sure why this line is needed either. Please explain.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+            // First convert all code blocks using <p> into <pre>
+            // which is used in almost all implementations of
+            // Markdown, even in the original Markdown specs.
+            formattedString = Regex.Replace(formattedString, @"<p><code>", "<pre><code>", RegexOptions.Multiline);
+            formattedString = Regex.Replace(formattedString, @"</code></p>", "</code></pre>", RegexOptions.Multiline);
+
+            System.Text.StringBuilder builder = new System.Text.StringBuilder();
+            System.Text.StringBuilder paragraphBuilder = new System.Text.StringBuilder();
+
+            var isNewLine = 0;
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: The variable name `isNewLine` doesn't really explain what this variable does.
How about something like `parseState`?

File: Journaley/Utilities/PostMarkdownParser.cs (hunk 13 above)
Comment (crxtrdude) [inline Journaley/Utilities/PostMarkdownParser.cs]: Yeah, I think you're right about that.

I actually assumed that problem, because supposedly in Day one you can do this:

```
~~Nikki~~
Nikki
~~Nikki~~
```

And it parses it correctly like:
~~Nikki~~
Nikki
~~Nikki~~

But in the previous code, it parsed like:

~~Nikki~~ Nikki
Nikki

That's why I added an regex check, but I think because of the foreach, it might have fixed it.

File: Journaley/Utilities/PostMarkdownParser.cs (hunk 14 above)
Comment (crxtrdude) [inline Journaley/Utilities/PostMarkdownParser.cs]: This was actually a proposed fix from a [MarkdownDeep PR](https://github.com/toptensoftware/markdowndeep/pull/22). I actually tested it on a regex tester, it does check for only line feeds in the end and at the same time disregards the end of the string.

In test:

```
<p>Line breaks are fun
Right?
It depends</p>
```

Only the first line and the next line are added with the `<br>` tag and not the last one.

File: Journaley/Utilities/PostMarkdownParser.cs (hunk 15 above)
Comment (crxtrdude) [inline Journaley/Utilities/PostMarkdownParser.cs]: The output in the end is actually a bit messy. Has a lot of line breaks. That's why I added a check for the carriage return + whitespace so that the output HTML string won't be messy.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+            StringBuilder builder = new System.Text.StringBuilder();
+            StringBuilder paragraphBuilder = new System.Text.StringBuilder();
+
+            // 0 - stumbles upon the beginning of a <p> tag/usual parsing.
+            // 1 - stumbles upon the end of </p> tag.
+            // 2 - stumbles upon a <pre> tag.
+            var parseState = 0;
+
+            string newLineString;
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: 1. Declare this variable when its used in line 86. This makes it easier to understand the code, because the scope of this variable is actually within the `else if (parseState == 0)` block. It is fine for the other variables like `builder`, `paragraphBuilder`, and `parseState`, because their state persists over the loops.
2. Let's rename this to `paragraph`, because it is a better name, it comes from the `paragraphBuilder`, and the name `newLineString` can be misleading because it usually refers to `\r\n` or some other platform dependent newline string.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+            string singleStrike;
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: Also declare this when its used.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+                    parseState = 0;
+                }
+                else if (line.Contains("<pre>"))
+                {
+                    parseState = -1;
+                }
+
+                if (parseState == 1)
+                {
+                    paragraphBuilder.AppendLine(line);
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: It's not appropriate to add this line here, because it won't be used and immediately cleared when this line also contains `</p>`. Let's add it in the `else` clause of the following `if` block.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+
+                    if (line.Contains("</p>"))
+                    {
+                        // Single line strikethroughs.
+                        singleStrike = Regex.Replace(line, @"(~~)(.*?)(~~)", "<del>$2</del>", RegexOptions.Singleline);
+                        builder.AppendLine(singleStrike);
+                        paragraphBuilder.Clear();
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: This line is no longer needed.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+                        parseState = 0;
+                    }
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs:46]: Add an `else` clause like this:

``` csharp
else
{
    paragraphBuilder.Add(line + "<br />");
}
```

I thought about this for a while, and I think this is the perfect place to add `<br />` tags.
This is the place where we're within the `<p>` tag, and seeing the line break before the `</p>` tag.
I don't think we need that complex regex below.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+                }
+                else if (parseState == 0)
+                {
+                    paragraphBuilder.AppendLine(line);
+
+                    newLineString = paragraphBuilder.ToString();
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: Please rename this variable to `paragraph` as I suggested above.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+
+                    // Multi line strikethrough
+                    newLineString = Regex.Replace(newLineString, @"(~~)(.*?)(~~)", "<del>$2</del>", RegexOptions.Multiline);
+
+                    // Single line breaks append with <br> tags.
+                    newLineString = Regex.Replace(newLineString, @"^([\w\*\>\<\[][^\r\n]*)(?=\r?\n[\w\*\>\<\[].*$)", "$1<br />", RegexOptions.Multiline);
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: Delete this code, now that we add the `<br />` tags above.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+                    paragraphBuilder.Clear();
+                }
+                else
+                {
+                    // For <pre> tagged text, just skip check and append to builder
+                    builder.AppendLine(line);
+                }
+            }
+
+            string strippedOverString = Regex.Replace(builder.ToString(), @"^(\r\s)", string.Empty, RegexOptions.Multiline);
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: Remove this code, because it has some unintended behavior. This also removes the whitespaces within the `<pre>` tag, which is definitely not what we want.

Having some whitespaces is not a big problem, and we can just leave them in there. The browser doesn't care about them anyways. Let's just return `builder.ToString()` here.

**Comment by yyoon** [comment]
Another issue I noticed: when playing with your test file, I found that the entry area shows nested scrollbars like this:

image

This happens because now we use the `<pre>` tag for the fenced code, and it doesn't wrap the line by default. We can fix this behavior by adding the CSS.

``` css
pre {
    word-wrap: break-word;
}
```

Please add this to all three css files (small, medium, large).

**Comment by crxtrdude** [comment]
Yeah this does not wrap the line by default, but the problem is not that, it's because the output of:

```
<code>` spans are delimited by backticks. You can include literal backticks like ``this``.
```

Which should result into this:

```
<p><code>&lt;code&gt;</code> spans are delimited by backticks. You can include literal backticks like <code>this</code>.</p>
```

Became this:

```
<pre><code>&lt;code&gt;</code> spans are delimited by backticks. You can include literal backticks like <code>this</code>.</p>
```

And the rest of the output became bad as a result as well. I need to be careful with this, I need to change my code to reflect that.

File: Journaley/Utilities/PostMarkdownParser.cs (hunk 21 above)
Comment (crxtrdude) [inline Journaley/Utilities/PostMarkdownParser.cs:46]: I think you thought this after I clarified to you about how the regex worked right?

**Comment by crxtrdude** [comment]
I have  fixed it, actually putting it on the parseState 0 so that it checks in a paragraph instead of a single line only.

Although actually, your commented fix on the `<br>` actually solved the problem of the `<p><code>` break. (Since it's enclosed in a `<p>` tag, even using a code block, `<br>` applies as one line break. `<pre>` on the other hand treats `<br>` as double)

Since people will not check the source of the HTML anyway, we could just omit that

**Comment by crxtrdude** [comment]
Sorry, pressed the wrong button

**Comment by yyoon** [comment]
Okay. That makes sense to me.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+        /// <param name="formattedString">Formatted HTML string</param>
+        /// <returns>Properly formatted HTML string for publishing.</returns>
+        public static string PostMarkdown(string formattedString)
+        {
+            StringBuilder builder = new StringBuilder();
+            StringBuilder paragraphBuilder = new StringBuilder();
+
+            // 0 - stumbles upon the beginning of a <p> tag/usual parsing.
+            // 1 - stumbles upon the end of </p> tag.
+            // 2 - skips check and just dumps the line to builder.
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: I think you meant to say -1 here, not 2.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+            var parseState = -1;
+
+            bool firstTitleParsed = false;
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: I'm sorry, but could you make the firstTitle feature as a separate PR, once this one is merged? This seems to require more thoughts on it.
Also, it seems like the current implementation would convert the first single line sentence into a title, even if it's not at the top.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+                else if (line.Contains("<pre>"))
+                {
+                    parseState = -1;
+                }
+
+                if (parseState == 1)
+                {
+                    if (line.Contains("</p>"))
+                    {
+                        string singleLine = line;
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: I think it would be fine to just use `line` in the following code. No need to introduce a new variable here.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+                        if (!firstTitleParsed)
+                        {
+                            singleLine = Regex.Replace(singleLine, @"(<p>)(.*?)(</p>)", "<h2>$2</h2>\n<br/>");
+                            firstTitleParsed = true;
+                        }
+
+                        builder.AppendLine(ReplaceMDStrikethrough(singleLine));
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: I don't think applying strikethrough here is correct. See my comments below.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 29]
```diff
@@ -0,0 +1,112 @@
...
+                        parseState = -1;
+                    }
+                    else
+                    {
+                        paragraphBuilder.AppendLine(ReplaceMDStrikethrough(line) + "<br />");
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: This one seems fine.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 30]
```diff
@@ -0,0 +1,112 @@
...
+                    }
+                }
+                else if (parseState == 0)
+                {
+                    paragraphBuilder.AppendLine(ReplaceMDStrikethrough(line));
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: Again, this seems incorrect to me.

I suggest applying strikethrough to the entire paragraph at this point, instead of applying it line by line.
That way, it could handle strikethroughs over multiple lines correctly.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 31]
```diff
@@ -0,0 +1,112 @@
...
+
+                    builder.AppendLine(paragraphBuilder.ToString());
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: Use `Append` instead of `AppendLine` to avoid adding extra new lines.

File: Journaley/Utilities/PostMarkdownParser.cs [hunk 32]
```diff
@@ -0,0 +1,112 @@
...
+        }
+        
+        /// <summary>
+        /// Replaces "~~" strikethrough Markdown tag into del HTML tags.
+        /// </summary>
+        /// <param name="parsedString">Any string that has strikethrough tags</param>
+        /// <returns>A string with del HTML tags in place of Markdown strikethrough tags.</returns>
+        private static string ReplaceMDStrikethrough(string parsedString)
+        {
+            parsedString = Regex.Replace(parsedString, @"(~~)(.*?)(~~)", "<del>$2</del>", RegexOptions.Singleline);
```
Comment (yyoon) [inline Journaley/Utilities/PostMarkdownParser.cs]: I think having just this one line with `Multiline` option would be enough, when we pass the entire paragraph to this method. Remove the following two lines.

File: Journaley/Utilities/PostMarkdownParser.cs (hunk 30 above)
Comment (crxtrdude) [inline Journaley/Utilities/PostMarkdownParser.cs]: I think you need to evaluate the private function.

**Comment by crxtrdude** [comment]
The reason for those two lines is because of something like this:
Supposed you got something like this:

```
~~Line 1
Line 2
Line 3~~
```

Obviously it's expected like this:

```
<p><del>Line 1
Line 2
Line 3<del><p>
```

Which is what GFM and Day One renders:
~~Line 1
Line 2
Line 3~~

For this kind of paragraph, that's why I added those 2 lines on this case.

**Comment by crxtrdude** [comment]
Why do I always press the first button ...

But anyway, I did agree with passing Strikethrough in paragraph, which did work on my side.

**Comment by yyoon** [comment]
I understand why you did that, but I still think it was incorrect for a few reasons.

First, it wouldn't necessarily match the opening / closing strikethrough mark pairs.
Take the following example.

``` markdown
~~Line1
Line2
Line3
```

GitHub renders it as:

~~Line1
Line2
Line3

, because there is no closing ~~ mark at the end.

However, the current code seems like it would just add the opening `<del>` and then never close it.
Processing the opening / closing mark separately will cause all sorts of this problems.

Second, it seems to me that it would only handle the ~~ marks that appear right after `<p>` and right before `</p>` am I right? This may not be able to handle the following case:

``` markdown
Normal text ~~strikethrough
should be applied here~~ and not here.
```

GitHub renders it as:

Normal text ~~strikethrough
should be applied here~~ and not here.

, but I think our previous code would probably not handle this correctly.

These are the reasons why I suggested to process at the paragraph level, not at the line level.

**Comment by crxtrdude** [comment]
The code I'm doing (not commited yet) is now processing it on the paragraph level now. I switched every line process over to paragraph level.

Your second example BTW, has not been rendered on Github, but it does render it on Day One and since we need to be as close to day one as possible, I might need to have a intelligent way of doing this.

**Comment by crxtrdude** [comment]
Why does the button for closing a PR is the one I always push? It's getting my nerves a bit.

**Comment by yyoon** [comment]
I guess space between `~~` and the text matters:

```
Normal text ~~strikethrough
should be applied here~~ and not here.
```

Normal text ~~strikethrough
should be applied here~~ and not here.

```
Normal text ~~ strikethrough
should be applied here~~ and not here.
```

Normal text ~~ strikethrough
should be applied here~~ and not here.

**Comment by crxtrdude** [comment]
~~I managed to actually successfully implement it. I will update our Test Syntax to reflect some of these changes. I'll commit it now.~~

**Comment by crxtrdude** [comment]
On second thought, there are problems that are going on. I'll fix this tomorrow morning.

**Comment by crxtrdude** [comment]
Here's the (hopefully last) implementation of PostMarkdown. I removed strikethrough support for now, because there are a lot of things to consider, but what I wanted to make sure was the other parts, _especially_ single line breaks, are implemented.

If we commit this now, with little to no changes, we'll just add the strikethrough problem as an issue and then work on it incrementally.

**Comment by crxtrdude** [comment]
Sorry, I need to do a followup for this one. Apparently, parsing backtick fenced code blocks post markdown would not escape HTML tags.

This is like similar to what I said (I think I said it) where `<br />` could be used inside a `<code>` block and it will render a newline.

That's why in this fix, instead of having to fight MarkdownDeep, I'll just convert the backtick fence into a tilde fence, that way it would be transformed properly as it is intended. The user's file of course is not affected by it since I only change it on UpdateWebBrowser only.

I'll find a way to do this programmatically on PostMarkdown properly, but the priority here is making the single line break work.

**Comment by yyoon** [comment]
Looks good to me. Thanks!
